drivers/misc/bcm2835_smi.c

```diff
@@ -0,0 +1,1108 @@
+	struct platform_device *pdev;
+
+	if (!node)
+		return NULL;
+	pdev = of_find_device_by_node(node);
+	if (!pdev)
+		return NULL;
+	inst = platform_get_drvdata(pdev);
+	if (!inst)
+		return NULL;
+
+	spin_lock_irqsave(&inst->register_lock, irq_flags);
+	if (!inst->device_registered) {
+		inst->device_registered = true;
+		rv = inst;
+	}
+	spin_unlock_irqrestore(&inst->register_lock, irq_flags);
+	return rv;
+}
+EXPORT_SYMBOL(bcm2835_smi_register);
+
+void bcm2835_smi_unregister(struct bcm2835_smi_instance *inst)
+{
+	unsigned long irq_flags;
+
+	if (inst) {
+		spin_lock_irqsave(&inst->register_lock, irq_flags);
+		inst->device_registered = false;
+		spin_unlock_irqrestore(&inst->register_lock, irq_flags);
+	}
+}
+EXPORT_SYMBOL(bcm2835_smi_unregister);
+
+/****************************************************************************
+*
+*   bcm2835_smi_probe - called when the driver is loaded.
+*
+***************************************************************************/
+
+static int bcm2835_smi_dma_setup(struct bcm2835_smi_instance *inst)
+{
+	dma_cap_mask_t mask;
+	int i, rv = 0;
+	bool alloc_failed = false;
+
+	dma_cap_zero(mask);
+	dma_cap_set(DMA_SLAVE, mask);
+	inst->dma_chan = dma_request_channel(mask, NULL, NULL);
+
+	inst->dma_config.src_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
+	inst->dma_config.dst_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
+	/* DREQ channel: SMI */
+	inst->dma_config.slave_id = 4 | (15 << 28) | (8 << 24);
+	inst->dma_config.src_addr = inst->smi_regs_busaddr
+		+ offsetof(struct smi_register_map, SMID);
+	inst->dma_config.dst_addr = inst->dma_config.src_addr;
+	/* Direction unimportant - always reconfigured before use */
+	inst->dma_config.direction = DMA_DEV_TO_MEM;
+	dmaengine_slave_config(inst->dma_chan, &inst->dma_config);
+	/* Alloc and map bounce buffers */
+	for (i = 0; i < DMA_BOUNCE_BUFFER_COUNT; ++i) {
+		inst->bounce.buffer[i] =
+		dma_alloc_coherent(inst->dev, DMA_BOUNCE_BUFFER_SIZE,
+				&inst->bounce.phys[i],
+				GFP_KERNEL);
+		if (!inst->bounce.buffer[i]) {
+			dev_err(inst->dev, "Could not allocate buffer!");
+			alloc_failed = true;
+			rv = -ENOMEM;
+			break;
+		}
+		inst->bounce.sgl[i] = smi_alloc_scatterlist_from_buffer
+		(
+			inst,
+			inst->bounce.phys[i],
+			DMA_BOUNCE_BUFFER_SIZE
+		);
+	}
+
+	if (alloc_failed) {
+		for (i = 0; i < DMA_BOUNCE_BUFFER_COUNT; ++i) {
+			if (!inst->bounce.buffer[i])
+				break;
+			dma_free_coherent(inst->dev, DMA_BOUNCE_BUFFER_SIZE,
+				inst->bounce.buffer[i], inst->bounce.phys[i]);
+			kfree(inst->bounce.sgl[i]);
+			inst->bounce.buffer[i] = NULL;
+			inst->bounce.sgl[i] = NULL;
+		}
+	}
+
+	return alloc_failed;
+}
+
+static int bcm2835_smi_probe(struct platform_device *pdev)
+{
+	int err;
+	struct device *dev = &pdev->dev;
+	struct device_node *node = dev->of_node;
+	struct resource *ioresource;
+	struct bcm2835_smi_instance *inst;
+
+	/* Allocate buffers and instance data */
+
+	inst = kzalloc(sizeof(struct bcm2835_smi_instance), GFP_KERNEL);
+
```

Comment:
Use devm_kzalloc() and freeing is automatic.
The same goes for inst->buffer_sgl